Accessibility parent

jzucker2 opened the pull request:
This is mostly a rough draft, wanted to see what you thought of this API for matching child elements.

jzucker2 committed:
Initial commit with SLAccessibilityContainer and working tests

jzucker2 committed:
Initial commit with SLAccessibilityContainer and working tests

jzucker2 committed:
Added SLTableViewCell subclass of SLAccessibilityContainer

jzucker2 committed:
Added SLTableViewCell subclass of SLAccessibilityContainer

jzucker2 committed:
Added test for TableViewCell methods

jzucker2 committed:
Added test for TableViewCell methods

jzucker2 committed:
Cleaned up comments

jzucker2 committed:
Cleaned up comments

jzucker2 commented:
I know it's missing documentation, if you like the approach, I'll add docs and resubmit. If not, we can figure out something a little different.

wearhere commented:
Thanks for the PR @jzucker2. Perfectly reasonable to submit a WIP. What would be helpful instead of docs though would be a little bit (here in comments) about the problem that the new API's solving. We've talked about it before but formalizing it here will help me understand the request and best comment on its direction.

jzucker2 commented:
This is a formal addition to Subliminal for matching and testing against elements inside a UITableViewCell. Subliminal does not have great abstraction for child and parent elements. This formalizes that in an easy API.

My big concern is how I fixed the slAccessibilityIsVisible method, by comparing against UITableTextAccessibilityElement, which isn't very clean, but it was an awkward comparison there. Really looking for feedback on that.

As well, this can be extended with a subclass for UICollectionViewCell, but I waited to do that until I heard about the UITableViewCell patch.

And technically, this API could be extended to any elements inside another view or accessibility container.

jzucker2 committed:
Updated SLAccessibilityContainer to be true subclass.

jzucker2 committed:
Updated SLAccessibilityContainer to be true subclass.

jzucker2 committed:
Cleaned up method names for table view cells

jzucker2 committed:
Cleaned up method names for table view cells

jzucker2 commented:
This is a cleaner implementation. The creation of SLAccessibilityContainer is a little awkward, but I had to get around initWithPredicate being a private method. This works though, except it struggles with isVisible because it seems to only select the mock accessibility view and not the actual view for isVisible. I fixed with a small hack but would love a stronger solution.

jzucker2 committed:
Cleaned up enum for AX container types

jzucker2 committed:
Cleaned up enum for AX container types

Commit:
Include collection view cell mock views in accessibility paths. (fixes #178)

Commit:
When confirming the visibility of unknown objects, don't wait to resolve the elements.

The superclass' implementation looks like:

    [self waitUntilTappable:NO
          thenPerformActionWithUIARepresentation:^(NSString *UIARepresentation) {
        isVisible = [[[SLTerminal sharedTerminal] evalWithFormat:@"%@.isVisible()", UIARepresentation] boolValue];
    } timeout:0.0];

"Unknown objects" == objects of unknown classes like the accessibility elements
vended by `UIWebBrowserView`.

We shouldn't wait to resolve the element because `-isVisible` is supposed
to evaluate the current state.

Commit:
Temporarily use UIAutomation to check visibility in non-portrait orientations. (refs #135)

As a workaround for the failure of our visibility routine.

Commit:
Ensure that `-[SLDevice setOrientation:]` waits for the device to rotate.

The previous delay was not long enough for the animation to complete on the iPad.

Commit:
Reset the device orientation after deactivation on iPads running iOS 5.1 or 7.1. (refs #180)

On iPads running iOS 5.1 or 7.1, `UIDevice` forgets its orientation after deactivation.
This is not only unexpected but can mess with `-[SLElement isVisible]`
which (as of #180) falls back upon `UIAutomation`'s visibility routine
if (but only if) the device is in a non-portrait orientation.

ahaneyinkling committed:
Merge pull request #186 from inkling/jeff/update_copyright_year

Update the copyright year.

ahaneyinkling committed:
Merge pull request #179 from inkling/jeff/collection_view_cells

Include collection view cell mock views in accessibility paths.

ahaneyinkling committed:
Merge pull request #180 from inkling/jeff/135_landscape_visibility

Workaround the non-portrait-orientation bug in our visibility routine.

jzucker2 committed:
Merge branch 'accessibilityParent' of https://github.com/jzucker2/Subliminal into accessibilityParent

jzucker2 commented on Sources/Classes/Internal/NSObject+SLVisibility.m:
```diff
@@ -133,7 +133,7 @@ - (BOOL)slAccessibilityIsVisible {
             // (thus is z-ordered before us/our parent)
             // and contains our hitpoint, it covers us
-            if (CGRectContainsPoint([element accessibilityFrame], testPoint)) return NO;
+            if (![NSStringFromClass([element class]) isEqualToString:@"UITableTextAccessibilityElement"] && CGRectContainsPoint([element accessibilityFrame], testPoint)) return NO;
```
This is the one issue I haven't been able to figure out for myself.

jzucker2 committed:
Removed focus from tests for official commit

jzucker2 commented:
The failure is from lack of documentation, not any compilation/test failures.

jzucker2 commented:
I'm closing this pull request in favor of a cleaner implementation that I just submitted: https://github.com/inkling/Subliminal/pull/190
